### PR TITLE
[SPARK-34890][PYTHON] Port/integrate Koalas main codes into PySpark

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

As a first step of [SPARK-34849](https://issues.apache.org/jira/browse/SPARK-34849), this PR proposes porting the Koalas main code into PySpark.

This PR contains minimal changes to the existing Koalas code as follows:
1. `databricks.koalas` -> `pyspark.pandas`
2. `from databricks import koalas as ks` -> `from pyspark import pandas as pp`
3. `ks.xxx -> pp.xxx`

Other than them:
1. Added a line to `python/mypy.ini` in order to ignore the mypy test. See related issue at [SPARK-34941](https://issues.apache.org/jira/browse/SPARK-34941).
2. Added a comment to several lines in several files to ignore the flake8 F401. See related issue at [SPARK-34943](https://issues.apache.org/jira/browse/SPARK-34943).

When this PR is merged, all the features that were previously used in [Koalas](https://github.com/databricks/koalas) will be available in PySpark as well.

Users can access to the pandas API in PySpark as below:

```python
>>> from pyspark import pandas as pp
>>> ppdf = pp.DataFrame({"A": [1, 2, 3], "B": [15, 20, 25]})
>>> ppdf
   A   B
0  1  15
1  2  20
2  3  25
```

The existing "options and settings" in Koalas are also available in the same way:

```python
>>> from pyspark.pandas.config import set_option, reset_option, get_option
>>> ppser1 = pp.Series([1, 2, 3])
>>> ppser2 = pp.Series([3, 4, 5])
>>> ppser1 + ppser2
Traceback (most recent call last):
...
ValueError: Cannot combine the series or dataframe because it comes from a different dataframe. In order to allow this operation, enable 'compute.ops_on_diff_frames' option.

>>> set_option("compute.ops_on_diff_frames", True)
>>> ppser1 + ppser2
0    4
1    6
2    8
dtype: int64
```

Please also refer to the [API Reference](https://koalas.readthedocs.io/en/latest/reference/index.html) and [Options and Settings](https://koalas.readthedocs.io/en/latest/user_guide/options.html) for more detail.

**NOTE** that this PR intentionally ports the main codes of Koalas first almost as are with minimal changes because:
- Koalas project is fairly large. Making some changes together for PySpark will make it difficult to review the individual change.
    Koalas dev includes multiple Spark committers who will review. By doing this, the committers will be able to more easily and effectively review and drive the development.
- Koalas tests and documentation require major changes to make it look great together with PySpark whereas main codes do not require.
- We lately froze the Koalas codebase, and plan to work together on the initial porting. By porting the main codes first as are, it unblocks the Koalas dev to work on other items in parallel.

I promise and will make sure on:
- Rename Koalas to PySpark pandas APIs and/or pandas-on-Spark accordingly in documentation, and the docstrings and comments in the main codes.
- Triage APIs to remove that don’t make sense when Koalas is in PySpark

The documentation changes will be tracked in [SPARK-34885](https://issues.apache.org/jira/browse/SPARK-34885), the test code changes will be tracked in [SPARK-34886](https://issues.apache.org/jira/browse/SPARK-34886).

### Why are the changes needed?

Please refer to:
- [[DISCUSS] Support pandas API layer on PySpark](http://apache-spark-developers-list.1001551.n3.nabble.com/DISCUSS-Support-pandas-API-layer-on-PySpark-td30945.html)
- [[VOTE] SPIP: Support pandas API layer on PySpark](http://apache-spark-developers-list.1001551.n3.nabble.com/VOTE-SPIP-Support-pandas-API-layer-on-PySpark-td30996.html)

### Does this PR introduce _any_ user-facing change?

Yes, now users can use the pandas APIs on Spark

### How was this patch tested?

Manually tested for exposed major APIs and options as described above.

### Koalas contributors

Koalas would not have been possible without the following contributors:

@ueshin 
@HyukjinKwon
@rxin
@xinrong-databricks
@RainFung
@charlesdong1991
@harupy
@floscha
@beobest2
@thunterdb
@garawalid
@LucasG0
@shril
@deepyaman
@gioa
@fwani
@90jam
@thoo
@AbdealiJK
@abishekganesh72
@gliptak
@DumbMachine
@dvgodoy
@stbof
@nitlev
@hjoo
@gatorsmile
@tomspur
@icexelloss
@awdavidson
@guyao
@akhilputhiry
@scook12
@patryk-oleniuk
@tracek
@dennyglee
@athena15
@gstaubli
@WeichenXu123
@hsubbaraj
@lfdversluis
@ktksq
@shengjh
@margaret-databricks
@LSturtew
@sllynn
@manuzhang
@jijosg
@sadikovi

### CODE REVIEW

File: python/pyspark/pandas/__init__.py
Comment (gengliangwang) [inline python/pyspark/pandas/__init__.py:2]: @itholic Please update the copyrights, thanks!

Comment (HyukjinKwon) [reply python/pyspark/pandas/__init__.py:2]: Oh yeah.

**Comment by HyukjinKwon** [comment]
cc @ueshin, @xinrong-databricks, @rxin from Koalas dev FYI.

It would be great if we have approvals from non-Koalas dev especially to make sure if your plan in PR description sounds making sense to everybody and non-Koalas dev people too. cc @holdenk @viirya @BryanCutler @zero323 FYI

File: python/pyspark/pandas/__init__.py
Comment (itholic) [reply python/pyspark/pandas/__init__.py:2]: Oh, thanks :)

**Comment by MaxGekk** [comment]
Is it possible to preserve commits history from the koalas repo (include all commits) otherwise you will hide the work of others contributors to the Koalas project.

**Comment by HyukjinKwon** [comment]
We could .. but that will include too many commits (https://github.com/databricks/koalas/graphs/contributors). I know people use that status for many other purposes, e.g.) some employees use that stats as some kind of metric of OSS contribution. I would like to minimize such potential impact.

One alternative is that to list up all authors as co-authors so single commit will have all other contributors as co-authors. However, that also might be counterintuitive because we're poring test codes and documentations separately. Especially test codes and docs might need multiple separate PRs with major changes. In this case, I think it makes less sense to keep all contributors.

**Comment by HyukjinKwon** [comment]
Probably we should at least include their GitHub IDs in the commit message. Let me address it.

File: python/pyspark/pandas/__init__.py
Comment (maropu) [inline python/pyspark/pandas/__init__.py:34]: We keep the name `Koalas` in the PySpark code?

Comment (HyukjinKwon) [reply python/pyspark/pandas/__init__.py:34]: Nope, we will replace it to either pandas-on-Spark and/or PySpark pandas APIs. They will be renamed right away after this PR gets merged (as explained in the PR description).

Comment (maropu) [reply python/pyspark/pandas/__init__.py:34]: Ah, okay, I missed that. Thanks.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by itholic** [comment]
retest this please

**Review by HyukjinKwon — APPROVED** [review]

**Review by zero323 — COMMENTED** [review]
Just thinking out loud ‒ wouldn't be better to have unambiguous package name (i.e. `pyspark.koalas` instead of `pyspark.pandas`)?

 I assume that in a long run we might  want to, for example,  integrate this feature with standard Spark configuration.  And since we already have pre-existing Pandas integration, having unrelated options referring to `pandas` could be confusing.

**Comment by HyukjinKwon** [comment]
Yeah, actually I have thought a lot about it, and discussed with some other people offline.

> since we already have pre-existing Pandas integration, having unrelated options referring to pandas could be confusing.

This is a very good point .. I am thinking about using a different name such as pandas-on-Spark internally. So, for example, we could have a configuration such as `spark.pandas-on-spark.blahblah`.

My current thought is to stick to use `pyspark.pandas` because:
- I checked few references such as Modin (which is probably the most similar case with us) that uses `modin.pandas`.
- Spark in general does not have its own naming in a component up to my best knowledge. As an example, [Shark](https://github.com/amplab/shark) became Spark SQL
- Koalas might not be a good name in a long run either (as far as I know it was more like related to branding?) - it might be best to clarify in its component name in a way.

However, I am open to change and to other names if many people think that `pyspark.koalas` or other alternatives are better.

**Comment by zero323** [comment]
It is not just about configuration ‒ in many ways this part of the API will be unique. With no feature parity and serious break from current API style its a thing of its own. So if I am for example interested in solution for some problem, I  am more likely to benefit from existing Koalas (or even Pandas) recipes, than general Spark. So having unique search term might be a nice thing.

Anyway... I just wanted to raise this in case other people have similar thoughts. One way or another I am happy to see it merged.

**Review by zero323 — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Thanks @zero323. I will leave this open for a couple of days more.

File: python/pyspark/pandas/__init__.py
Comment (viirya) [inline python/pyspark/pandas/__init__.py:63]: This looks like for checking PySpark version in Koalas.  Do we still need to check PySpark version as it is now included in PySpark?

**Review by viirya — APPROVED** [review]
I'm fine with the approach described in the PR.

File: python/pyspark/pandas/__init__.py
Comment (itholic) [reply python/pyspark/pandas/__init__.py:63]: Correct!
We'll address the all codes related with the existing third-party dependencies in the SPARK-34887.
Thanks for the comment!

**Comment by garawalid** [comment]
@HyukjinKwon @zero323 
I agree with both of you, the name of the API should be unique.
I'd like to point out that using dashes in the name API will force us to import the package with `import_module`:
```python
# from pyspark.pandas-on-spark import api -> It doesn't work
api = import_module('pyspark.pandas-on-spark.api')
```
Also, we should follow the [PEP8 module naming convention](https://www.python.org/dev/peps/pep-0008/#package-and-module-names).

**Review by garawalid — COMMENTED** [review]
LGTM

**Comment by HyukjinKwon** [comment]
Yeah, thanks for pointing that out. I actually more meant `spark.pandas-on-spark.blahblah` by the configuration name in Spark, e.g.) `spark.set("spark.pandas-on-spark.blahblah", "true")`. Since Koalas has its own option system, we won't likely have to deal with it for now but you never know what's going to happen in the future though ..

Speaking of `pyspark.pandas` and `pyspark.koalas`, I understand it's difficult to choose one over another as each has distinct limitations and benefits - there was a hot debate on this offline too.

Again, thanks for your contribution in Koalas @garawalid. This PR would have never been possible without you.

**Comment by lfdversluis** [comment]
I think the change makes sense.
Does this change also mean that the CI/CD pipelines will be merged? Is there any impact there like longer test runtimes or code style checks?

**Comment by itholic** [comment]
Thanks for the comment, @lfdversluis .
Yes, we will also migrate all Koalas unit tests and static analysis in SPARK-34886 which may increase the PySpark testing time.
We will try to minimize the runtime it in the process of integrating with the current PySpark test.

**Comment by HyukjinKwon** [comment]
Merged to master.

Thanks all!

**Pull request closed without merging**
